tests/test1.cpp

```diff
@@ -0,0 +1,113 @@
+#include <iostream>
+#include "uast.h"
+
+class Node
+{
+public:
+  Node(std::string i) : internal_type(i) {}
+  std::string internal_type;
+  std::string token;
+
+  std::vector<Node *> children;
+  std::vector<uint16_t> roles;
+
+  void add_child(Node *node)
+  {
+    children.push_back(node);
+  }
+
+  void add_role(uint16_t role)
+  {
+    roles.push_back(role);
+  }
+
+  std::string as_string()
+  {
+    return internal_type;
+  }
+};
+
+const char *
+get_internal_type(const void *node)
+{
+  return ((Node *)node)->internal_type.data();
+}
+
+const char *
+get_token(const void *node)
+{
+  return ((Node *)node)->token.data();
+}
+
+int get_children_size(const void *node)
+{
+  return ((Node *)node)->children.size();
+}
+
+void *get_child(const void *node, int index)
+{
+  return ((Node *)node)->children.at(index);
+}
+
+int get_roles_size(const void *node)
+{
+  return ((Node *)node)->roles.size();
+}
+
+uint16_t get_role(const void *node, int index)
+{
+  return ((Node *)node)->roles.at(index);
+}
+
+UAST get_c_api()
+{
+  return UAST(NodeImpl{
```

Comment:
As you said, this probably needs another name like NodeAPI or something like that.